model-optimizer/extensions/ops/sparse_reshape.py

```diff
@@ -28,26 +29,77 @@ def supported_attrs(self):
 
     @staticmethod
     def infer(node: Node):
+        name = node.soft_get('name', node.id)
+
         input_indices_shape = node.in_port(0).data.get_shape()
         input_indices_value = node.in_port(0).data.get_value()
-        input_shape_value = node.in_port(1).data.get_value()
-        new_shape_value = node.in_port(2).data.get_value()
+        input_shape = node.in_port(1).data.get_value()
+        new_shape = node.in_port(2).data.get_value()
         new_shape_shape = node.in_port(2).data.get_shape()
 
-        assert input_shape_value is not None and new_shape_value is not None, \
+        assert input_shape is not None and new_shape is not None, \
             "Values for input shape and new shape must be defined"
-        assert np.count_nonzero(new_shape_value == -1) <= 1, \
+        assert np.count_nonzero(new_shape == -1) <= 1, \
             "Value -1 occurs in new shape value more than once"
 
-        node.out_port(1).data.set_shape(new_shape_shape)
-        output_shape_value = new_shape_value
-        if np.count_nonzero(output_shape_value == -1) == 1:
-            elem = np.prod(input_shape_value) // np.prod(new_shape_value[new_shape_value != -1])
-            output_shape_value[output_shape_value == -1] = elem
-        node.out_port(1).data.set_value(output_shape_value)
+        num_of_input_elements = np.prod(input_shape)
+        num_of_output_elements = 1
+        for index, x in enumerate(new_shape):
+            if x is dynamic_dimension:
+                num_of_output_elements = dynamic_dimension
+                break
+            elif x == 0 and node.has_and_set('special_zero'):
+                if input_shape[index] is not dynamic_dimension:
+                    num_of_output_elements *= input_shape[index]
+            elif x != -1:
+                num_of_output_elements *= x
+
+        # input_shape = [dyn, 5, 6], new_shape = [0, -1] => output_shape [dyn, 30]
```

Comment:
this is not correct. output_shape must be `[0, dyn]`

If we have any `dyn` in `input_shape` and `-1` in new_shape then `-1` will be substituted with `dyn`.
If we don't have any `-1` `input_shape` and we have only `-1` in `new_shape` then this `-1` can be deduced in `output_shape`. In all other cases `-1` should be substituted with `dyn`

Reply:
I checked in TF and yes, in the case below it leaved `output_shape` of uncertain dimension without 30. 

The reason I added this statement: I believed there will no harm to concretize `output_shape`, since there is `0` in `output_shape` there is no way that zeroth dimension size of `output_shape` will be different from `input_shape`. Therefore we can deduce -1 as product of left dimensions.

Shapes calculated by us do not contradict with TF shapes they are just more strict. But I can remove this code.

Reply:
changed according to your suggestions